Sources/PackageExtension/PackageExtension.swift

```diff
@@ -0,0 +1,339 @@
+    /// package extension itself may also write to this directory.
+    public let outputDir: Path
+    
+    /// A directory into which the package extension or the tool it invokes can
+    /// write any caches that speed up its operation or any other intermediate
+    /// files that shouldn't be further processed.
+    public let cacheDir: Path
+    
+    /// Looks up and returns the path of a named command line executable tool.
+    /// The executable must be either in the toolchain or in the system search
+    /// path for executables, or be provided by an executable target or binary
+    /// target on which the package extension target depends. Returns nil, but
+    /// does not throw an error, if the tool isn't found. Extensions that re-
+    /// quire the tool should emit an error diagnostic if it cannot be found.
+    public func lookupTool(named name: String) throws -> Path {
+        // TODO: Rather than just appending the name, this should instead use
+        // a mapping of tool names to paths (passed in from the context).
+        return execsDir.appending(name)
+    }
+    
+    /// A directory in which the built executables available to the extension
+    /// will be located.
+    /// TODO: This should instead be a mapping of tool names to paths (passed
+    /// in from the context).
+    private let execsDir: Path
+}
+
+/// Constructs commands to run during the build, including full command lines.
+/// All paths should be based on the ones passed to the extension in the target
+/// build context.
+public final class CommandConstructor {
+    
+    fileprivate init() {}
+    
+    /// Creates a command to run during the build. The executable should be a
+    /// path returned by `TargetBuildContext.lookupTool(named:)`, and all the
+    /// paths in both the command line and the input and output lists should be
+    /// based on the paths in the target build context structure.
+    ///
+    /// Note that input and output dependencies are ignored for prebuild and
+    /// postbuild actions, since they always run before and after the build
+    /// respectively.
+    public func createCommand(displayName: String?, executable: Path, arguments: [String], workingDirectory: Path? = nil, environment: [String: String]? = nil, inputPaths: [Path] = [], outputPaths: [Path] = [], derivedSourcePaths: [Path] = []) {
+        output.commands.append(Command(executable: executable, arguments: arguments, workingDirectory: workingDirectory, environment: environment, displayName: displayName, inputPaths: inputPaths, outputPaths: outputPaths, derivedSourcePaths: derivedSourcePaths))
+    }
+    
+    func addGeneratedOutputFile(path: Path) {
+        output.generatedFilePaths.append(path.string)
+    }
+    
+    public func addPrebuildOutputDirectory(path: Path) {
+        output.prebuildOutputDirectories.append(path.string)
+    }
+}
+
+/// A constructed command.
+fileprivate struct Command: Encodable {
```

Comment:
side question: why do all these need to be decodable / coddle?

Reply:
These are passed back to SwiftPM via JSON.  That's also how the context comes in.

Reply:
This part is just like PackageDescription.  But it's not visible in the public-facing API that that's what happens.